ompi/mca/coll/base/coll_base_alltoall.c

```diff
@@ -296,43 +379,102 @@ int ompi_coll_base_alltoall_intra_bruck(const void *sbuf, size_t scount,
         }
     }
 
+    for (distance = 1; distance < size; distance *= radix) {
+        if (distance <= size) {
+            largest_power_of_k = distance;
+        }
+    }
+
+    tmp_sbuf = (void **)malloc(sizeof(void *) * (radix - 1));
+    tmp_rbuf = (void **)malloc(sizeof(void *) * (radix - 1));
+    for (i = 0; i < radix - 1; i++) {
+        tmp_sbuf[i] = (void *) malloc(rext * rcount * largest_power_of_k);
+        tmp_rbuf[i] = (void *) malloc(rext * rcount * largest_power_of_k);
+    }
+
+    max_reqs = 2 * (radix - 1);
+    reqs = ompi_coll_base_comm_get_reqs(module->base_data, max_reqs);
+
     /* perform communication step */
-    for (distance = 1; distance < size; distance<<=1) {
+    for (distance = 1; distance < size; distance *= radix) {
+        num_reqs = 0;
+        for (i = 1; i < radix; i++) {
 
-        sendto = (rank + distance) % size;
-        recvfrom = (rank - distance + size) % size;
+            if (distance * i >= size) {
+                break;
+            }
 
-        new_ddt = ompi_datatype_create((1 + size/distance) * (2 + rdtype->super.desc.used));
+            sendto = (rank + distance * i) % size;
+            recvfrom = (rank - distance * i + size) % size;
+
+            /* first offset where the phase'th bit has value i */
+            offset = distance * i;
+            /* number of consecutive occurrences of i */
+            nconsecutive_occurrences = distance;
+            /* distance between non-consecutive occurrences of i */
+            delta = (radix - 1) * distance;
+            packsize = 0;
+
+            /* Pack non-contiguous data from tmpbuf to tmp_sbuf */
+            while (offset < size) {
```

Comment:
We can save extra memcpy on L463 by directly copying data received in the previous round `tmp_rbuf` to `tmp_sbuf`.
```c
if (i > 1) {
  // Copy from tmp_rbuf
} else {
  // Copy from tmpbuf <--- this can be further optimized away by using original sbuf
}
```